superbench/benchmarks/micro_benchmarks/ib_validation_performance/ib_validation_performance.cc

```diff
@@ -0,0 +1,450 @@
+    MPI_Bcast(ports.data(), pair_count, MPI_INT, 0, MPI_COMM_WORLD);
+    return ports;
+}
+
+// Get and gather ips for all ranks
+void gather_ips(vector<string> &ips) {
+    // get my ip
+    string ip = get_my_ip();
+    // convert my ip address from string to vector<char>
+    int my_len = ip.size();
+    int max_len = 16;
+    char my_str_padded[max_len + 1];
+    memset(my_str_padded, '\0', max_len + 1);
+    memcpy(my_str_padded, ip.c_str(), my_len);
+    // prepare buffer to save ips from all ranks
+    char *all_str = NULL;
+    int all_len = (max_len + 1) * world_size;
+    all_str = (char *)malloc(all_len * sizeof(char));
+    memset(all_str, '\0', all_len);
+
+    // gather ips from all ranks
+    MPI_Allgather(my_str_padded, max_len + 1, MPI_CHAR, all_str, max_len + 1, MPI_CHAR, MPI_COMM_WORLD);
+
+    // restore all ips from char* to vector<string>
+    char *str_idx = all_str;
+    for (int rank_idx = 0; rank_idx < world_size; rank_idx++) {
+        ips.push_back(str_idx);
+        str_idx = str_idx + max_len + 1;
+    }
+}
+
+// Parse raw output of ib command
+float process_raw_output(string output) {
+    float res = -1.0;
+    try {
+        vector<string> lines;
+        split_string(output, lines, "\n");
+        regex re("\\d+\\s+\\d+\\s+\\d+\\.\\d+\\s+(\\d+\\.\\d+)\\s+\\d+\\.\\d+");
+        for (string line : lines) {
+            smatch m;
+            if (regex_search(line, m, re))
+                res = std::max(res, stof(m.str(1)));
+        }
+    } catch (const std::exception &e) {
+        std::cout << "Error: failed to parse raw_output: " << output << std::endl;
+    }
+
+    return res;
+}
+
+// Run ib command on server/client with server ip
+float run_cmd(string cmd_prefix, int port, bool server, string ip) {
+    string is_server = server ? "server" : "client";
+    // client sleep 1s in case that client starts before server
+    if (!server) {
+        std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+    }
+
+    string command, output;
+    try {
+        // exec command in termimal
+        command = cmd_prefix + " -p " + to_string(port);
+        command = server ? command : command + " " + ip;
+        output = exec(command.c_str());
+    } catch (const std::exception &e) {
+        std::cout << "Error: failed to exec command: " << command << std::endl;
+    }
+
+    // parse raw output and get result
+    float result = process_raw_output(output);
+
+    return result;
+}
+
+// The ranks in vector of (server, client) run commands parallel
+vector<float> run_cmd_parallel(string cmd_prefix, const vector<std::pair<int, int>> &run_pairs_in_parallel,
+                               const vector<int> &ports, const vector<string> &ips) {
+    int size = run_pairs_in_parallel.size();
+    // invoke function to run cmd in multi threads mode for each rank in the pairs
+    unordered_map<int, std::future<float>> threads;
+    for (int index = 0; index < size; index++) {
```

Comment:
should we limit max number of active threads? (given that some clusters can be really large, and also with multiple NICs per node)